packages/vats/src/orchestration.js

```diff
@@ -0,0 +1,255 @@
+        const { onOpen } = this.state;
+        if (onOpen && typeof onOpen === 'function') {
+          void onOpen(connection, localAddr, remoteAddr);
+        }
+      },
+      async onClose(connection, reason) {
+        trace(`ICA Channel closed. Reason: ${reason}`);
+        const { onClose } = this.state;
+        if (onClose && typeof onClose === 'function') {
+          void onClose(connection, reason);
+        }
+      },
+      async onReceive(connection, bytes) {
+        const { onReceive } = this.state;
+        if (onReceive && typeof onReceive === 'function') {
+          void onReceive(connection, bytes);
+        }
+        return '';
+      },
+    },
+  );
+
+export const OrchestrationI = M.interface('Orchestration', {
+  provideAccount: M.callWhen(M.string(), M.string())
+    .optional(M.remotable('Port'))
+    .returns(M.remotable('ChainAccount')),
+  getChain: M.callWhen(M.string()).returns(M.remotable('Chain')),
+});
+
+/**
+ * @param {import('@agoric/base-zone').Zone} zone
+ * @param {ReturnType<typeof prepareChainAccount>} createChainAccount
+ * @param {ReturnType<typeof prepareConnectionHandler>} makeConnectionHandler
+ */
+const prepareOrchestration = (
+  zone,
+  createChainAccount,
+  makeConnectionHandler,
+) =>
+  zone.exoClassKit(
+    'Orchestration',
+    {
+      admin: M.interface('OrchestrationAdmin', {
+        foo: M.call().returns(M.string()),
+      }),
+      public: OrchestrationI,
+    },
+    /** @param {Partial<OrchestrationPowers>} [initialPowers] */
+    initialPowers => {
+      /** @type {PowerStore} */
+      const powers = zone.detached().mapStore('PowerStore');
+      if (initialPowers) {
+        for (const [name, power] of Object.entries(initialPowers)) {
+          powers.init(/** @type {keyof OrchestrationPowers} */ (name), power);
+        }
+      }
+      return { powers, icaControllerNonce: 0 };
+    },
+    {
+      admin: {
+        foo() {
+          return 'bar';
+        },
+      },
+      public: {
+        /**
+         * @param {string} hostConnectionId the counterparty connection_id
+         * @param {string} controllerConnectionId self connection_id
+         * @param {Port} [currentPort] if a port is provided, it will be reused
+         *   to create the account
+         */
+        async provideAccount(
+          hostConnectionId,
+          controllerConnectionId,
+          currentPort,
+        ) {
+          await null;
+          const connectionHandler = makeConnectionHandler();
+          let port;
+          if (!matches(currentPort, M.remotable('Port'))) {
+            if (currentPort) trace('Invalid Port provided, binding a new one.');
+            const network = getPower(this.state.powers, 'network');
+
+            port = await E(network)
+              .bind(`/ibc-port/icacontroller-${this.state.icaControllerNonce}`)
+              .catch(e => Fail`Failed to bind port: ${bare(e)}`);
```

Comment:
does this throw away e's stack? use error cause instead?

```suggestion
              .catch(e => assert.fail(`Failed to bind port`, { cause: e }));
```